subsys/bluetooth/controller/flash_driver.c

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <errno.h>
+
+#include <zephyr.h>
+#include <device.h>
+#include <soc.h>
+#include <flash.h>
+
+#include "ble_controller_soc.h"
+#include "multithreading_lock.h"
+
+#include <logging/log.h>
+#define LOG_MODULE_NAME blectrl_flash
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL_DBG);
+
+/* TODO: Shouldn't this macro be defined in some kernel header? Doesn't seems
+ * like it is available globally. */
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+#define FLASH_DRIVER_WRITE_BLOCK_SIZE 1
+
+static struct {
+	/** Used to ensure a single ongoing operation at any time.  */
+	struct k_sem sem;
+	const void *data;
+	off_t addr;
+	u16_t len;
+	u16_t prev_len;
+	uint32_t tmp_word;      /**< Used for unalinged writes. */
+	/* NOTE: Read is not async, so not a part of this enum. */
+	enum {
+		FLASH_OP_NONE,
+		FLASH_OP_WRITE,
+		FLASH_OP_ERASE
+	} op;
+} flash_state;
+
+/* Forward declarations */
+static int btctlr_flash_read(struct device *dev, off_t offset, void *data, size_t len);
+static int btctlr_flash_write(struct device *dev, off_t offset, const void *data, size_t len);
+static int btctlr_flash_erase(struct device *dev, off_t offset, size_t size);
+static int btctlr_flash_write_protection_set(struct device *dev, bool enable);
+
+static int flash_op_execute(void);
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static void btctlr_flash_page_layout_get(struct device *dev,
+					 const struct flash_pages_layout **layout,
+					 size_t *layout_size);
+#endif /* defined(CONFIG_FLASH_PAGE_LAYOUT) */
+
+static const struct flash_driver_api btctrl_flash_api = {
+	.read = btctlr_flash_read,
+	.write = btctlr_flash_write,
+	.erase = btctlr_flash_erase,
+	.write_protection = btctlr_flash_write_protection_set,
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	.page_layout = btctlr_flash_page_layout_get,
+#endif  /* CONFIG_FLASH_PAGE_LAYOUT */
+	.write_block_size = FLASH_DRIVER_WRITE_BLOCK_SIZE
+};
+
+
+/* Utility functions */
+static inline bool is_addr_valid(off_t addr, size_t len)
+{
+	if ((addr + len) > NRF_FICR->CODEPAGESIZE * NRF_FICR->CODESIZE ||
```

Comment:
You may use `dev_layout `here to speed up this check

Reply:
I would need to surround the block with `CONFIG_FLASH_PAGE_LAYOUT` or hide it behind a macro. Do you think the speed gain reading RAM vs FICR is worth it? It's only retrieving this once every API call, writing/erasing is a much larger hit.

Reply:
I guess you can leave it as is